crypto/evp/pmeth_lib.c

```diff
@@ -104,7 +107,8 @@ static EVP_PKEY_CTX *int_ctx_new(EVP_PKEY *pkey, ENGINE *e, int id)
     if (id == -1) {
         if (!pkey || !pkey->ameth)
             return NULL;
-        id = pkey->ameth->pkey_id;
+
+        id = pkey->type;
```

Comment:
While you are at it, could you replace first logical `!` with explicit NULL check and omit second one. If also appropriate to add empty line  prior outermost if... Actually one can wonder if one needs the NULL check at all... I mean if subroutine ends up being called with pkey being NULL and id being -1, then can we view it as programmer's mistake (or memory corruption)? Because if we can, then it would be appropriate to omit the check.